File: validate-subsequence/ValidateSubsequence.js
function isValidSubsequence1(array, sequence) {
    let count = 0;
    for (let index = 0; index < array.length; index++) {
        if(sequence.length === count) break;
        if(array[index] === sequence[count]) { count++; }
    }
    return sequence.length == count;
  }

  function isValidSubsequence2(array, sequence) {
    let count = sequence.length - 1;
    for(let index = 0; index < array.length; index++) {
        if(count === -1) break;
        if(array[array.length - 1 - index ] === sequence[count]) {
            count--;
        }
    }
    return count === -1;
  }

  function isValidSubsequence3(array, sequence) {
    for(let index = 0; index < array.length; index++) {
        if(sequence.length === 0) break
        if(array[array.length - 1 - index ] === sequence[sequence.length - 1]) {
           sequence.pop();
        }
    }
    return sequence.length === 0;
  }

  function isValidSubsequence4(array, sequence) {
    for(let index = 0; index < array.length; index++) {
        if(sequence.length === 0) break
        if(array[index] === sequence[0]) {
           sequence.shift();
        }
    }
    return sequence.length === 0;
  }

  exports.isValidSubsequence = isValidSubsequence4;

File: validate-subsequence/ValidateSubsequence.java
import java.util.*;

public class ValidateSubsequence {
    public static boolean isValidSubsequence1(List<Integer> array, List<Integer> sequence) {
        int count = 0;
        int index = 0;
        while(index < array.size()) {
            if(sequence.size() == count) break;
            if(array.get(index).equals(sequence.get(count))) { count++; }
            index++;
        }
        return sequence.size() == count;
    }

    public static boolean isValidSubsequence2(List<Integer> array, List<Integer> sequence) {
        int count = sequence.size() - 1;
        int index = array.size() - 1;
        while(index >= 0) {
            if(count == -1) break;
            if(array.get(index).equals(sequence.get(count))){ count--; }
            index--;
        }
        return count == -1;
    }

    public static boolean isValidSubsequence3(List<Integer> array, List<Integer> sequence) {
        Deque<Integer> numbers = new ArrayDeque<>(sequence);
        int index = 0;
        while(index < array.size()) {
            if(numbers.isEmpty()) break; 
            if(array.get(index).equals(numbers.getFirst())) { numbers.removeFirst(); }
          index++;
        }
        return numbers.isEmpty();
    }

    public static boolean isValidSubsequence4(List<Integer> array, List<Integer> sequence) {
        Deque<Integer> numbers = new ArrayDeque<>(sequence);
        int index = array.size() - 1;
        while(index >= 0) {
            if(numbers.isEmpty()) break;
            if(array.get(index).equals(numbers.getLast())) { numbers.removeLast(); }
            index--;
        }
        return numbers.isEmpty();
    }

    public static boolean isValidSubsequence5(List<Integer> array, List<Integer> sequence) {
        Stack<Integer> stack = new Stack<>(); stack.addAll(sequence);
        int index = 0;
        while(index < array.size()) {
            if(stack.isEmpty()) break;
            if(array.get(array.size() - index - 1).equals(stack.peek())) { stack.pop(); }
            index++;
        }
        return stack.isEmpty();
    }
}
